### ISSUE TITLE
Use multi-row inserts for massive speedups on to_sql over high latency connections

### ISSUE THREAD

**Issue by maxgrenderjones** [pandas-dev/pandas#8953]
I have been trying to insert ~30k rows into a mysql database using pandas-0.15.1, oursql-0.9.3.1 and sqlalchemy-0.9.4. Because the machine is as across the atlantic from me, calling `data.to_sql` was taking >1 hr to insert the data. On inspecting with wireshark, the issue is that it is sending an insert for every row, then waiting for the ACK before sending the next, and, long story short, the ping times are killing me.

However, following the instructions from [SQLAlchemy](http://docs.sqlalchemy.org/en/latest/core/dml.html#sqlalchemy.sql.expression.Insert.values), I changed 

``` python
def _execute_insert(self, conn, keys, data_iter):
    data = [dict((k, v) for k, v in zip(keys, row)) for row in data_iter]
    conn.execute(self.insert_statement(), data)
```

to 

``` python
def _execute_insert(self, conn, keys, data_iter):
    data = [dict((k, v) for k, v in zip(keys, row)) for row in data_iter]
    conn.execute(self.insert_statement().values(data))
```

and the entire operation completes in less than a minute. (To save you a click, the difference is between multiple calls to `insert into foo (columns) values (rowX)` and one massive `insert into foo (columns) VALUES (row1), (row2), row3)`). Given how often people are likely to use pandas to insert large volumes of data, this feels like a huge win that would be great to be included more widely.

Some challenges:
- Not every database supports multirow inserts (SQLite and SQLServer didn't in the past, though they do now). I don't know how to check for this via SQLAlchemy
- The MySQL server I was using didn't allow me to insert the data all in one go, I had to set the chunksize (5k worked fine, but I guess the full 30k was too much). If we made this the default insert, most people would have to add a chunk size (which might be hard to calculate, as it might be determined by the maximum packet size of the server).

The easiest way to do this, would be to add a `multirow=` boolean parameter (default `False`) to the `to_sql` function, and then leave the user responsible for setting the chunksize, but perhaps there's a better way?

Thoughts?

**Comment by nhockham** [comment]
We've figured out how to monkey patch - might be useful to someone else. Have this code before importing pandas.

```
from pandas.io.sql import SQLTable

def _execute_insert(self, conn, keys, data_iter):
    print "Using monkey-patched _execute_insert"
    data = [dict((k, v) for k, v in zip(keys, row)) for row in data_iter]
    conn.execute(self.insert_statement().values(data))

SQLTable._execute_insert = _execute_insert

```

**Comment by jorisvandenbossche** [comment]
Maybe we can just start with adding this feature through a new `multirow=True` keyword (with a default of False for now), and then we can later always see if we can enable it by default?

@maxgrenderjones @nhockham interested to do a PR to add this?

**Comment by mangecoeur** [comment]
@jorisvandenbossche I think it's risky to start adding keyword arguments to address specific performance profiles. If you can guarantee that it's faster in all cases (if necessary by having it determine the best method based on the inputs) then you don't need a flag at all.

Different DB-setups may have different performance optimizations (different DB perf profiles, local vs network, big memory vs fast SSD, etc, etc), if you start adding keyword flags for each it becomes a mess.

I would suggest creating subclasses of SQLDatabase and SQLTable to address performance specific implementations, they would be used through the object-oriented API. Perhaps a "backend switching" method could be added but frankly using the OO api is very simple so this is probably overkill for what is already a specialized use-case.

I created such a sub-class for loading large datasets to Postgres (it's actually much faster to save data to CSV then use the built-in non-standard COPY FROM sql commands than to use inserts, see https://gist.github.com/mangecoeur/1fbd63d4758c2ba0c470#file-pandas_postgres-py). To use it you just do `PgSQLDatabase(engine, <args>).to_sql(frame, name,<kwargs>)`

**Comment by artemyk** [comment]
Just for reference,  I tried running the code by @jorisvandenbossche (Dec 3rd post) using the multirow feature.  It's quite a bit slower.  So the speed-tradeoffs here is not trivial:

```
In [4]: engine = create_engine('sqlite:///:memory:') #, echo='debug')

In [5]: df = pd.DataFrame(np.random.randn(50000, 10))

In [6]: 

In [6]: %timeit df.to_sql('test_default', engine, if_exists='replace')
1 loops, best of 3: 1.05 s per loop

In [7]: 

In [7]: from pandas.io.sql import SQLTable

In [8]: 

In [8]: def _execute_insert(self, conn, keys, data_iter):
   ...:         data = [dict((k, v) for k, v in zip(keys, row)) for row in data_iter]
   ...:         conn.execute(self.insert_statement().values(data))
   ...:     

In [9]: SQLTable._execute_insert = _execute_insert

In [10]: 

In [10]: reload(pd)
Out[10]: <module 'pandas' from '/usr/local/lib/python2.7/site-packages/pandas/__init__.pyc'>

In [11]: 

In [11]: %timeit df.to_sql('test_default', engine, if_exists='replace', chunksize=10)
1 loops, best of 3: 9.9 s per loop
```

Also, I agree that adding keyword parameters is risky.  However, the multirow feature seems pretty fundamental.  Also, 'monkey-patching' is probably not more robust to API changes than keyword parameters.

**Comment by mangecoeur** [comment]
Its as i suspected. Monkey patching isn't the solution I was suggesting - rather that we ship a number of performance oriented subclasses that the informed user could use through the OO interface (to avoid loading the functional api with too many options)

-----Original Message-----

**Comment by maxgrenderjones** [comment]
As per the initial ticket title, I don't think this approach is going to be preferable in all cases, so I wouldn't make it the default. However, without it, the pandas `to_sql` unusable for me, so it's important enough for me to continue to request the change. (It's also become the first thing I change when I upgrade my pandas version). As for sensible `chunksize` values, I don't think there is one true `n`, as the packet size will depend on how many columns there are (and what's in them) in hard to predict ways. Unfortunately SQLServer fails with an error message that looks totally unrelated (but isn't) if you set the `chunksize` too high (which is probably why multirow inserts aren't turned on except with a patch in SQLAlchemy), but it works fine with `mysql`. Users may need to experiment to determine what value of `n` is likely to result in an acceptably large packet size (for whatever their backing database is). Having pandas chose `n` is likely to get land us way further down in the implementation details than we want to be (i.e. the opposite direction from the maximum-possible-abstraction SQLALchemy approach)

In short, my recommendation would be to add it as a keyword, with some helpful commentary about how to use it. This wouldn't be the first time a keyword was used to select an implementation (see: http://pandas.pydata.org/pandas-docs/dev/generated/pandas.DataFrame.apply.html) but that perhaps isn't the best example, as I haven't the first idea about what `raw=` means, even having read the explanation!

**Comment by dragonator4** [comment]
I have noticed that it also consumes a huge amount of memory. Like a 1.6+ GB DataFrame with some 700,000 rows and 301 columns requires almost 34 GB during insert! That is like over the top inefficient. Any ideas on why that might be the case? Here is a screen clip:

![image](https://cloud.githubusercontent.com/assets/8196363/17013812/02b51276-4ede-11e6-8006-9fd32cb01e14.png)

**Comment by andreacassioli** [comment]
Hi guys,
any progress on this issue?

I am try to insert around 200K rows using to_sql but it takes forever and consume a huge amount of memory! Using chuncksize helps with the memory but still the speed is very slow. 

My impression, looking at the MSSQL DBase trace is that the insertion is actually performed one row at the time. 



The only viable approach now is to dump to a csv file on a shared folder and use BULK INSERT. But it very annoying and inelegant!

**Comment by ostrokach** [comment]
@andreacassioli You can use `odo` to insert a DataFrame into an SQL database through an intermediary CSV file. See [Loading CSVs into SQL Databases](http://odo.pydata.org/en/latest/perf.html). I don't think there is a way to come even close to `BULK INSERT` performance using ODBC.

**Comment by andreacassioli** [comment]
@ostrokach thank you, indeed I am using csv files now. If I could get close, I would trade a bit of time for simplicity!

**Comment by indera** [comment]
I thought this might help somebody:
http://docs.sqlalchemy.org/en/latest/faq/performance.html#i-m-inserting-400-000-rows-with-the-orm-and-it-s-really-slow

**Comment by jorisvandenbossche** [comment]
@indera pandas does not use the ORM, only sqlalchemy Core (which is what the doc entry there suggests to use for large inserts)

**Comment by russlamb** [comment]
is there any consensus on how to work around this in the meantime?  I'm inserting a several million rows into postgres and it takes forever.  Is CSV / odo the way to go?

**Comment by jreback** [comment]
@russlamb a practical way to solve this problem is simply to bulk upload. This is someone db specific though, so ``odo`` has solutions for ``postgresl`` (and may be ``mysql``) I think. for something like sqlserver you have to 'do this yourself' (IOW you have to write it).

**Comment by indera** [comment]
For sqlserver I used the approach

**Comment by jorisvandenbossche** [comment]
>  Is CSV / odo the way to go?

This solution will almost always be faster I think, regardless of the multi-row / chunksize settings. 

But, @russlamb, it is always interesting to hear whether such a multi-row keyword would be an improvement in your case. See eg https://github.com/pandas-dev/pandas/issues/8953#issuecomment-76139975 on a way to easily test this out.

I think there is agreement that we want to have a way to specify this (without necessarily changing the default). So if somebody wants to make a PR for this, that is certainly welcome. 
There was only some discussion on how to add this ability (new keyword vs subclass using OO api).

**Comment by indera** [comment]
@jorisvandenbossche The document I linked above mentions "Alternatively, the SQLAlchemy ORM offers the Bulk Operations suite of methods, which provide hooks into subsections of the unit of work process in order to emit Core-level INSERT and UPDATE constructs with a small degree of ORM-based automation."

What I am suggesting is to implement a sqlserver specific version for "to_sql" which under the hood uses the SQLAlchemy core for speedups.

**Comment by mangecoeur** [comment]
This was proposed before. The way you go is to implement an pandas sql
class optimised for a backend. I posted a gist in the past for using
postgres COPY FROM command which is much faster. However something similar
is now available in odo, and built in a more robust way. There isn't much
point IMHO in duplicating work from odo.

**Comment by mangecoeur** [comment]
Also noticed you mentioned sqlalchemy could core instead. Unless something
has changed a lot, only sqlalchemy core is used in any case, no orm. If you
want to speed up more than using core you have to go to lower level, db
specific optimisation

**Comment by dfernan** [comment]
Is this getting fixed/taken care of? As of now inserting pandas dataframes into a SQL db is extremely slow unless it's a toy dataframe.  Let's decide on a solution and push it forward?

**Comment by ostrokach** [comment]
@dfernan As mentioned above, you may want to look at the [odo](http://odo.pydata.org/en/latest/perf.html). Using an intermediary CSV file will always be orders of magnitude faster that going through sqlalchemy, no matter what kind of improvements happen here...

**Comment by markschwarz** [comment]
@ostrokach, I'm unconvinced that odo's behavior is what the typical Pandas user wants.  

To speak for myself, I just spent a few hours switching from the monkey patch above to odo.  Plain pandas runtime was 10+ hours, RBAR.  The monkey patch runs in 2 on the same data set. 
 The odo/CSV route was faster, as expected, but not by enough to make it worth the effort.  I fiddled with CSV conversion issues I didn't care much about, all in the name of avoiding the monkey patch.   I'm importing 250K rows from ~10 mysql and PG DBs into a common area in Postgres, for NLP analysis.  

I'm intimately familiar with the bulk loading approaches odo espouses.  I've used them for years where I'm starting with CSV data.  There are key limitations to them:

1. For the df->CSV->Postgres case, shell access and a scp step is needed to get the CSV on the PG host.  It looks like @mangecoeur has gotten around this with a stream to STDIN.
2. For my purpose (~1M rows of comments, with lots of special cases in the text content) I struggled to get CSV parameters right.  I didn't want performance gains badly enough to keep investing in this.

I switch backed the the patch, so I could get back to my analysis work.  

I agree with @jorisvandenbossche, @maxgrenderjones.  An option (not a default) to choose this would be immensely useful.  @artemyk's point about dialect.supports_multivalues_insert might make the default faster, too.

I'm glad to submit a PR if that'd move this forward.

**Comment by s-trooper** [comment]
just to add my experience with odo, it has not worked for MS Sql bulk inserts because of known issue with encoding.  imho m-row insert are good practical solution for most ppl.

**Comment by jreback** [comment]
@markschwarz an option to enable this to work faster would be very welcome!

**Comment by makmanalp** [comment]
Tracing the queries using sqlite, I do seem to be getting in multi-inserts when using `chunksize`:

```
2017-09-28 00:21:39,007 INFO sqlalchemy.engine.base.Engine INSERT INTO country_hsproduct_year (location_id, product_id, year, export_rca, import_value, cog, export_value, distance, location_level, product_level) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
2017-09-28 00:21:39,007 INFO sqlalchemy.engine.base.Engine ((75, 1237, 1996, 1.7283086776733398, 273487116.0, 0.0, 514320160.0, 0.5413745641708374, 'country', '4digit'), (75, 1237, 1997, 1.7167805433273315, 312047528.0, 0.0, 592372864.0, 0.5314807891845703, 'country', '4digit'), (75, 1237, 1998, 1.2120152711868286, 341676961.0, 0.0, 468860608.0, 0.5472233295440674, 'country', '4digit'), (75, 1237, 1999, 1.236651062965393, 334604240.0, 0.0, 440722336.0, 0.5695921182632446, 'country', '4digit'), (75, 1237, 2000, 1.189828872680664, 383555023.0, 0.0, 426384832.0, 0.5794379711151123, 'country', '4digit'), (75, 1237, 2001, 0.9920380115509033, 374157144.0, 0.3462945520877838, 327031392.0, 0.6234743595123291, 'country', '4digit'), (75, 1237, 2002, 1.0405025482177734, 471456583.0, 0.0, 377909376.0, 0.6023964285850525, 'country', '4digit'), (75, 1237, 2003, 1.147829532623291, 552441401.0, 0.0, 481313504.0, 0.5896202325820923, 'country', '4digit')  ... displaying 10 of 100000 total bound parameter sets ...  (79, 1024, 2015, 0.0, None, 0.8785018920898438, 0.0, 0.9823430776596069, 'country', '4digit'), (79, 1025, 1995, 0.0, None, 0.5624096989631653, 0.0, 0.9839603304862976, 'country', '4digit'))
```

**Comment by makmanalp** [comment]
(without the monkey patch, that is)

**Comment by makmanalp** [comment]
Interestingly, with the monkey patch, it breaks when i give it chunksize of 10^5, but not 10^3. The error is "too many sql variables" on sqlite.

**Comment by markschwarz** [comment]
@makmanalp, I haven't traced on PG to check for that behavior, yet, but I nearly always set chunksize on insert.  In my example above, I randomly set it to 5 values between 200-5000.  I didn't see drastic elapsed time differences among those choices, without the monkey patch.  With the patch, elapsed time dropped ~80%.

**Comment by makmanalp** [comment]
@hangyao I think that patch is implementation-specific, one of those things that python DBAPI leaves to the DBAPI driver on how to handle. So it could be faster or not.  RE: the syntax error, unsure about that.

**Comment by tsktsktsk123** [comment]
I am thinking about adding the function below in the `/io/sql.py` file on line 507 within the `_engine_builder` function within a new IF clouse on line 521 making the 'new' `_engine_builder` the below snippet. I've tested it briefly on my environment and it works great for MSSQL databases, achieving >100x speed-ups. I haven't tested it on other databases yet.

The thing from witholding me to make a PR is that I think it's more effort to make it neat and safe than just inserting it like below, this might not always be the desired spec and adding a boolean switch, which turns this setting on/off , (e.g. `fast_executymany=True`) in the `to_sql` seemed like a bit too large endeavor to just do without asking, I reckon.

So my questions are:

- Does the below function work and also increase the INSERT speed for PostgreSQL? 

- Does pandas event want this snippet in their source? If so:

- Is it desired to add this function to the default `sql.py` functionality or is there a better place to add this? 



Love to hear some comments.

Source for the answer: https://stackoverflow.com/questions/48006551/speeding-up-pandas-dataframe-to-sql-with-fast-executemany-of-pyodbc/48861231#48861231 

```
@event.listens_for(engine, 'before_cursor_execute')
def receive_before_cursor_execute(conn, cursor, statement, params, context, executemany):
    if executemany:
        cursor.fast_executemany = True
``` 
```
def _engine_builder(con):
    """
    Returns a SQLAlchemy engine from a URI (if con is a string)
    else it just return con without modifying it.
    """
    global _SQLALCHEMY_INSTALLED
    if isinstance(con, string_types):
        try:
            import sqlalchemy
        except ImportError:
            _SQLALCHEMY_INSTALLED = False
        else:
            con = sqlalchemy.create_engine(con)

    @event.listens_for(engine, 'before_cursor_execute')
    def receive_before_cursor_execute(conn, cursor, statement, params, context, executemany):
        if executemany:
            cursor.fast_executemany = True
return con
```

**Comment by jorisvandenbossche** [comment]
@tsktsktsk123 There has recently been a PR merged related to this: https://github.com/pandas-dev/pandas/pull/19664. I didn't yet look in detail to your post, and it's certainly not exactly the same (it uses `supports_multivalues_insert` attribute of the sqlalchemy engine), but just to make sure you are aware of it in case this already helps as well.

**Comment by tsktsktsk123** [comment]
That's great news! I haven't looked into the PR  but will compare it this weekend and get back with the results. Thanks for the heads up.

**Comment by grisaitis** [comment]
@schettino72 How much data were you inserting?

**Comment by schettino72** [comment]
Around 30K rows with 10 columns. But really almost anything I try is slower (SQL is faster but overall slower). It is creating a huge SQL statement where there is value interpolation for EVERY value. Something like 

```
 %(user_id_m32639)s, %(event_id_m32639)s, %(colx_m32639)s,
```

**Comment by citynorman** [comment]
I found `d6tstack` much simpler to use, it's a one-liner `d6tstack.utils.pd_to_psql(df, cfg_uri_psql, 'benchmark', if_exists='replace')`, see https://github.com/d6t/d6tstack/blob/master/examples-sql.ipynb

**Comment by VincentLa14** [comment]
I've been using the Monkey Patch Solution for some time now, but now I'm getting an error:

```
TypeError: insert_statement() missing 2 required positional arguments: 'data' and 'conn'
```

Is anyone else getting this?

**Comment by hnazkani** [comment]
is this getting fixed ? Currently, df.to_sql is extremely slow and can't be used at all for many practical use cases. Odo project seems to have been abandoned already.
I have following use cases in financial time series where df.to_sql is pretty much not usable:
1) copying historical csv data to postgres database  - can't use df.to_sql and had to go with custom code around psycopg2 copy_from functionality
2) streaming data (coming in a batch of ~500-3000 rows per second)  to be dumped to postgres database - again df.to_sql performance is pretty disappointing as it is taking too much time to insert these natural batches of data to postgres.
The only place where I find df.to_sql useful now is to create tables automatically !!! - which is not the use case it was designed for. 
I am not sure if other people also share the same concern but this issue needs some attention for "dataframes-to-database" interfaces to work smoothly.
Look forward.

**Comment by jconstanzo** [comment]
Hey, I'm getting this error when I try to perform a multi-insert to a SQLite database:

This is my code:
`df.to_sql("financial_data", con=conn, if_exists="append", index=False, method="multi")`

and I get this error:
```
Traceback (most recent call last):

  File "<ipython-input-11-cf095145b980>", line 1, in <module>
    handler.insert_financial_data_from_df(data, "GOOG")

  File "C:\Users\user01\Documents\Code\FinancialHandler.py", line 110, in insert_financial_data_from_df
    df.to_sql("financial_data", con=conn, if_exists="append", index=False, method="multi")

  File "C:\Users\user01\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\generic.py", line 2531, in to_sql
    dtype=dtype, method=method)

  File "C:\Users\user01\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\io\sql.py", line 460, in to_sql
    chunksize=chunksize, dtype=dtype, method=method)

  File "C:\Users\user01\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\io\sql.py", line 1547, in to_sql
    table.insert(chunksize, method)

  File "C:\Users\user01\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\io\sql.py", line 686, in insert
    exec_insert(conn, keys, chunk_iter)

  File "C:\Users\user01\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\io\sql.py", line 609, in _execute_insert_multi
    conn.execute(self.table.insert(data))

TypeError: insert() takes exactly 2 arguments (1 given)
```

Why is this happening? I'm using Python 3.7.3 (Anaconda), pandas 0.24.2 and sqlite3 2.6.0.

Thank you very much in advance!

**Comment by jorisvandenbossche** [comment]
@jconstanzo can you open this as a new issue? 
And if possible, can you try to provide a reproducible example? (eg a small example dataframe that can show the problem)

**Comment by Bauxitedev** [comment]
@jconstanzo  Having the same issue here. Using `method='multi'` (in my case, in combination with `chunksize`) seems to trigger this error when you try to insert into a SQLite database.

Unfortunately I can't really provide an example dataframe because my dataset is huge, that's the reason I'm using `method` and `chunksize` in the first place.

**Comment by jconstanzo** [comment]
I'm sorry for the delay. I just opened an issue for this problem: https://github.com/pandas-dev/pandas/issues/29921

**Comment by ban04toufuonline** [comment]
how to hacked this?
> Now I just need to persuade the SQLAlchemy folks to set `supports_multivalues_insert` to true on SQL Server >2005 (I hacked it into the code and it works fine, but it's not on by default).
> 
> On a more on-topic note, I think the chunksize could be tricky. On my mysql setup (which I probably configured to allow large packets), I can set chunksize=5000, on my SQLServer setup, 500 was too large, but 100 worked fine. However, it's probably true that most of the benefits from this technique come from going from inserting 1 row at a time to 100, rather than 100 to 1000.

**Comment by ban04toufuonline** [comment]
tow to hack  this? @maxgrenderjones 
> Now I just need to persuade the SQLAlchemy folks to set `supports_multivalues_insert` to true on SQL Server >2005 (I hacked it into the code and it works fine, but it's not on by default).
> 
> On a more on-topic note, I think the chunksize could be tricky. On my mysql setup (which I probably configured to allow large packets), I can set chunksize=5000, on my SQLServer setup, 500 was too large, but 100 worked fine. However, it's probably true that most of the benefits from this technique come from going from inserting 1 row at a time to 100, rather than 100 to 1000.
